File: npm/src/commands/cancel.ts
import { CLIContext } from '../cli';
import { apiPost } from '../lib/api';
import { readConfig, resolveEndpoint } from '../lib/config';
import { success, error, info, json as jsonOut } from '../lib/output';
import { confirm } from '../lib/prompt';

export async function cancel(ctx: CLIContext): Promise<void> {
  const config = readConfig();
  if (!config) {
    error('No .formhandle config found. Run "formhandle init" first.');
    process.exit(1);
  }

  const { domain, endpoint } = resolveEndpoint(config, ctx.flags.domain);

  if (!ctx.flags.json) {
    const yes = await confirm(`Cancel subscription for ${domain} (${endpoint.handler_id})?`);
    if (!yes) {
      info('Aborted.');
      return;
    }
  }

  const res = await apiPost(`/cancel/${endpoint.handler_id}`, {});

  if (res.status !== 200) {
    if (ctx.flags.json) {
      jsonOut({ error: res.data.error || res.data.message || 'Cancel failed', status: res.status });
    } else {
      error(String(res.data.error || res.data.message || 'Cancel failed'));
    }
    process.exit(1);
  }

  if (ctx.flags.json) {
    jsonOut({ ok: true, handler_id: endpoint.handler_id, message: res.data.message });
  } else {
    success(String(res.data.message || 'Check your email to confirm cancellation.'));
  }
}

File: python/src/formhandle/commands/cancel.py
"""Cancel subscription."""

import sys

from formhandle.api import api_post
from formhandle.config import read_config, resolve_endpoint
from formhandle.output import success, error, info, json_out
from formhandle.prompt import confirm


def run(args) -> None:
    config = read_config()
    if not config:
        error('No .formhandle config found. Run "formhandle init" first.')
        sys.exit(1)

    resolved = resolve_endpoint(config, getattr(args, "domain", None))
    domain = resolved["domain"]
    endpoint = resolved["endpoint"]

    if not args.json:
        if not confirm(f"Cancel subscription for {domain} ({endpoint['handler_id']})?"):
            info("Aborted.")
            return

    res = api_post(f"/cancel/{endpoint['handler_id']}", {})

    if args.json:
        if res["status"] == 200:
            json_out({"ok": True, "handler_id": endpoint["handler_id"], "message": res["data"].get("message", "")})
        else:
            json_out({"error": res["data"].get("error", "Cancel failed"), "status": res["status"]})
            sys.exit(1)
    else:
        if res["status"] == 200:
            success(res["data"].get("message", "Check your email to confirm cancellation."))
        else:
            error(res["data"].get("error", f"Cancel failed (HTTP {res['status']})"))
            sys.exit(1)
